src/site/en/blog/2020-07-09-no-to-routers/index.md

```diff
@@ -0,0 +1,246 @@
+</figure>
+
+There is a stable <abbr title="User experience">UX</abbr> trend to split long forms into multiple steps called "form wizards". If you are aiming for the good UX, you would usually prefer to add those steps to browser history. This will allow user to navigate between form steps with browser buttons.
+
+Sounds like a natural feature, right? But you can't implement it with basic router functionality. You will need to push those steps to browser history manually. Router will still listen to history events, so you will need to find a workaround to ignore them. Or you could just specify service paths, that are accessible only if the wizard is in a relevant state. Meaning users will share links to various wizard steps, and you will also need to check if they are allowed to go there directly.
+
+You would presume routers don't support this, because there are some browser limitations. Well, actually there is none: you can push the next history state without route change, but with another data.
+
+Routers also don't allow storing modal dialog state. It leads to hacks, like modals having their own paths or managing modals without router at all.
+
+And some routers even forbid mixing [pathnames](https://developer.mozilla.org/en-US/docs/Web/API/URL/pathname) and [hashes](https://developer.mozilla.org/en-US/docs/Web/API/URL/hash), while fragments (defined by hashes) are the vital part of the web platform and are the perfect fit for the use case of virtual routing. Which literally is rendering of different fragments of the same page depending on current data state.
+
+Yes, you guessed it correctly: virtual routing cannot be implemented because routers assume you only have plain mapping (see #2).
+
+# What should we do then?
+
+All this sounds awful. What should we do to solve our routing problem in a simple, maintainable and scalable way without going crazy in the mean time? I'm glad you asked.
+
+1. You still need to deserialise state from path and serialise it back. There are many libraries that do it just fine, you can check, for instance: [path-to-regexp](https://github.com/pillarjs/path-to-regexp) (used by react-router and page.js), [path-parser](https://github.com/troch/path-parser) (used by router5) or [route-parser](https://github.com/rcs/route-parser). You also need to put that state into browser history (or sometimes not), and there is [history.js](https://github.com/browserstate/history.js/) lib for you. But you should probably just use [Browser History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API) – it is in a good and consistent state across all major browsers right now.
+2. You need to add asynchronous dependencies check which consists of additional script loading and data requests. Usually, this is what your controller is already handling: redux-thunk, redux-saga, mobx, effector – whatever you prefer. If it handles your async logic, it could handle your async dependencies management.
```

Comment:
```suggestion
2. You need to add asynchronous dependencies check, which consists of additional script loading and data requests. Usually, your controller already handles it: redux-thunk, redux-saga, mobx, effector – whatever you prefer. If it handles your async logic, it could handle your async dependencies management.
```